analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/LoopDetector.cs

```diff
@@ -91,15 +91,16 @@ private static IEnumerable<int> Successors(ControlFlowGraph cfg, int index)
         {
             successors = successors.Concat(tryRegion.ReachableHandlers().Select(x => x.FirstBlockOrdinal));
         }
-        if (block.EnclosingNonLocalLifetimeRegion() is { Kind: ControlFlowRegionKind.Finally } finallyRegion)
+        if (block.EnclosingRegion(ControlFlowRegionKind.Finally) is { } finallyRegion
+            && block.Successors.Length == 1
+            && block.Successors[0].Semantics == ControlFlowBranchSemantics.StructuredExceptionHandling)
         {
             var tryFinallyRegion = finallyRegion.EnclosingRegion;
             var associatedTryRegion = tryFinallyRegion.NestedRegion(ControlFlowRegionKind.Try);
             successors = successors.Concat(
-                cfg.Blocks.Where(x => x.IsIn(associatedTryRegion))
-                    .SelectMany(x => x.SuccessorBlocks)
-                    .Where(x => !x.IsIn(tryFinallyRegion))
-                    .Select(x => x.Ordinal));
+                cfg.Blocks.SelectMany(x => x.Successors)
+                    .Where(x => x.Destination is not null && x.LeavingRegions.Contains(tryFinallyRegion))
+                    .Select(x => x.Destination.Ordinal));
```

Comment:
This change is unrelated to the logic of this PR. I just noticed that the LeavingRegions property exists and thought this is easier to read. Lmk if you see any problems with it (performance maybe?).